server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java

```diff
@@ -162,6 +164,44 @@ public void testRepositoriesStatsCanHaveTheSameNameAndDifferentTypeOverTime() {
         assertThat(repositoryStatsTypeB.getRepositoryStats(), equalTo(MeteredRepositoryTypeB.STATS));
     }
 
+    // this can happen when the repository plugin is removed, but repository is still exist
+    public void testHandlesUnknownRepositoryTypeWhenApplyingClusterState() {
+        var repoName = randomAlphaOfLengthBetween(10, 25);
+
+        var clusterState = createClusterStateWithRepo(repoName, "unknown");
+        repositoriesService.applyClusterState(new ClusterChangedEvent("starting", clusterState, emptyState()));
+
+        var repo = repositoriesService.repository(repoName);
+        assertThat(repo, isA(MissingPluginRepository.class));
```

Comment:
Given that this is a user-facing exception, I like to have some assertions about the message itself rather than just the exception type.